### ISSUE TITLE
Map to array of string, cannot save the data.

### ISSUE THREAD

**Issue by winnochan** [typegoose/typegoose#682]
## Versions

- System: linux
- NodeJS: 16.13.1
- Typescript: 4.5.2

- Compiler / Transpiler: ts-node (10.4.0)
- Typegoose(NPM): 9.7.0
- mongoose: 6.2.6 and 6.2.8
- mongodb: 5.0.6

## What is the Problem?

I defined a data ( Types.Map<string[]> ), but there is an error when setting and saving the document.

## Code Example

```ts
import { defaultClasses, getModelForClass, modelOptions, mongoose, prop } from '@typegoose/typegoose';
import { Types } from 'mongoose';

@modelOptions({ schemaOptions: { versionKey: false } })
export class Test implements defaultClasses.Base<string> {
  @prop()
  _id: string;

  get id() {
    return this._id;
  }

  @prop({ _id: false, default: {}, type: [String] })
  mapArr: Types.Map<string[]>;
}

const TestModel = getModelForClass(Test);

async function testCase() {
  try {
    await mongoose.connect('mongodb://localhost:27017');
    const _id = Date.now().toString();
    let doc = await TestModel.create({ _id });
    await doc.save();
    doc = await TestModel.findById(_id);

    doc.mapArr.set('key', ['arr', 'str']);

    console.log(_id, doc.modifiedPaths());
    await doc.save();
  } catch (err) {
    console.log(err);
  }

  await mongoose.disconnect();
}

testCase();
```

## Do you know *why* it happens?

*no*

---

## Error Message

```
1648019555013 []
Error: Test validation failed: mapArr.key: Cast to string failed for value "[ 'arr', 'str' ]" (type Array) at path "mapArr.$*"
    at ValidationError.inspect (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/error/validation.js:48:26)
    at formatValue (node:internal/util/inspect:763:19)
    at inspect (node:internal/util/inspect:340:10)
    at formatWithOptionsInternal (node:internal/util/inspect:2006:40)
    at formatWithOptions (node:internal/util/inspect:1888:10)
    at console.value (node:internal/console/constructor:323:14)
    at console.log (node:internal/console/constructor:359:61)
    at testCase (/home/winnochan/Projects/works/svrapp/test/test-mongoose-map2arr.ts:32:13)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  errors: {
    'mapArr.key': CastError: Cast to string failed for value "[ 'arr', 'str' ]" (type Array) at path "mapArr.$*"
        at SchemaString.cast (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/schema/string.js:603:11)
        at SchemaString.SchemaType.applySetters (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/schematype.js:1180:12)
        at Map.set (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/types/map.js:89:11)
        at testCase (/home/winnochan/Projects/works/svrapp/test/test-mongoose-map2arr.ts:27:16)
        at processTicksAndRejections (node:internal/process/task_queues:96:5) {
      stringValue: `"[ 'arr', 'str' ]"`,
      messageFormat: undefined,
      kind: 'string',
      value: [Array],
      path: 'mapArr.$*',
      reason: null,
      valueType: 'Array'
    }
  },
  _message: 'Test validation failed'
}
```

**Comment by hasezoey** [comment]
just to confirm, you are currently trying to make a Map of a array of strings (`Map<string, [string]>`), correct?

if that is actually the case, i dont quite know if this is actually supported by typegoose (i dont think we have tests for that), could you maybe provide [Debug Output](https://typegoose.github.io/typegoose/docs/guides/advanced/logger#enable-debug-logger)?

also the current workaround would be to make a intermediate class that holds the array:
```ts
class Intermediate {
  @prop({ type: [String] })
  public i?: string[];
}

class Parent {
  @prop({ type: () => Intermediate })
  public mapArr?: mongoose.Types.Map<mongoose.Types.Subdocument & Intermediate>; // but i think just "Intermediate" would also work
}
```

**Comment by winnochan** [comment]
@hasezoey Yes, Map<string, string[]>. This is the document link, [Map<string, string[]> Case](https://typegoose.github.io/typegoose/docs/api/decorators/prop#map-options).
![微信截图_20220323181318](https://user-images.githubusercontent.com/8348908/159679097-b98601b5-11ee-454b-b4b1-58b792b08c47.png)
This is DEBUG output.
```
Added "Test._id" to the Decorator Cache
Added "Test.mapArr" to the Decorator Cache
Assigning global Schema Options to "Test"
buildSchema called for "Test"
_buildSchema Called for Test with options: { versionKey: false }
Detecting PropType
Starting to process "Test._id"
Detecting PropType
Starting to process "Test.mapArr"
Prop Option "type" is set to  [ [Function: String] ]
Final getType: dim: 1, type: [Function: String]
Detecting PropType
mapOptions called
Converting "String" to mongoose Type
Final mapped Options for Type "String" { inner: {}, outer: { _id: false, default: {}, dim: 1 } }
Class "Test" already existed in the constructors Map
1648031071776 []
Error: Test validation failed: mapArr.key: Cast to string failed for value "[ 'arr', 'str' ]" (type Array) at path "mapArr.$*"
    at ValidationError.inspect (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/error/validation.js:48:26)
    at formatValue (node:internal/util/inspect:763:19)
    at inspect (node:internal/util/inspect:340:10)
    at formatWithOptionsInternal (node:internal/util/inspect:2006:40)
    at formatWithOptions (node:internal/util/inspect:1888:10)
    at console.value (node:internal/console/constructor:323:14)
    at console.log (node:internal/console/constructor:359:61)
    at testCase (/home/winnochan/Projects/works/svrapp/test/test-mongoose-map2arr.ts:34:13)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  errors: {
    'mapArr.key': CastError: Cast to string failed for value "[ 'arr', 'str' ]" (type Array) at path "mapArr.$*"
        at SchemaString.cast (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/schema/string.js:603:11)
        at SchemaString.SchemaType.applySetters (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/schematype.js:1180:12)
        at Map.set (/home/winnochan/Projects/works/svrapp/node_modules/mongoose/lib/types/map.js:89:11)
        at testCase (/home/winnochan/Projects/works/svrapp/test/test-mongoose-map2arr.ts:29:16)
        at processTicksAndRejections (node:internal/process/task_queues:96:5) {
      stringValue: `"[ 'arr', 'str' ]"`,
      messageFormat: undefined,
      kind: 'string',
      value: [Array],
      path: 'mapArr.$*',
      reason: null,
      valueType: 'Array'
    }
  },
  _message: 'Test validation failed'
}
```

**Comment by hasezoey** [comment]
can reproduce:
```ts
// NodeJS: 17.6.0
// MongoDB: 4.2-bionic (Docker)
import { getModelForClass, prop } from '@typegoose/typegoose'; // @typegoose/typegoose@9.7.0
import * as mongoose from 'mongoose'; // mongoose@6.2.4

class Main {
  @prop({ type: () => [String] })
  public mapArr?: Map<string, string[]>;
}

const MainModel = getModelForClass(Main);

(async () => {
  await mongoose.connect(`mongodb://localhost:27017/`, {
    dbName: 'verifyMASTER',
  });

  console.log('schema path', MainModel.schema.path('mapArr'));

  const doc = await MainModel.create({ mapArr: [['key1', ['val1']]] });

  console.log('doc', doc);

  await mongoose.disconnect();
})();
```

Err: `ValidationError: Main validation failed: mapArr.key1: Cast to string failed for value "[ 'val1' ]" (type Array) at path "mapArr.$*"`

Current Result: `mapArr` is in runtime of type `Map<string, string>`
Expected Result: `mapArr` to be in runtime of type `Map<string, string[]>`

---

> Yes, Map<string, string[]>. This is the document link, [Map<string, string[]> Case](https://typegoose.github.io/typegoose/docs/api/decorators/prop#map-options).

thanks for providing a link, i forgot this was in the documentation, but yes it currently does not work, but i think there are no tests for this

**Comment by hasezoey** [comment]
this issue should be fixed with 36deb80cd9a63542eb16c0a0fcb5db84ce5db81e, which will be in version 9.7.1, which will release after the builds are complete (commonly in around ~10 minutes)
